### ISSUE TITLE
Revamp slipstreaming

### ISSUE THREAD

**Issue by Alayan-stk-2** [supertuxkart/stk-code#3149]
Slipstreaming implies following another kart from close behind to get a speed boost.

This is currently an unbalanced feature ; for a few reasons : 
1)The speed bonus is significant and often unexpected. This can be destabilizing and push a player to go out of track (or have similar issues)
2)There is no visual indicator of slipstreaming progress. This plays a huge part for the destabilizing aspect, as the "is slipstreamed" cue doesn't tell us if we have slipstreamed for long enough.
3)It takes a long time to get the bonus (2 seconds minimum). It is hard enough to stay close enough to the AI kart (or more exactly, inside the right angle cone) to get it. This is especially an issue if there is any kind of turning, as even if we are in fact following the same trajectory, the direction in which the kart ahead is going will make the slipstream cone change.

![slipstreamangle](https://user-images.githubusercontent.com/25536748/37554302-ac409fb6-29d6-11e8-82c5-84cd50ce3d15.png)

4)The force of the bonus and the minimum speed required are independent of the difficulty level. It would work better if this was not the case (should help with karts getting it too easily right at the start).
5)It is risky enough to go close behind another karts with bowling balls/bubble gums in play.

After improvements, slipstreaming has the potential to be a skill-selecting feature - helping a better player to have a better place despite item randomness.

It could also help to reduce slightly the advantage of being first ahead by making the pack faster.

The AI already has code to use slipstream. The hardest part is already done.

Fixes : 
- [x] Add some kind of slipstream progress bar, with clear visual cue indicating when the player will get a bonus. The progress bar should be intersected by a minimum slipstream line and go further up to a maximum time at which no additional bonus is added. The progress bar may be displayed only when the kart is slipstreaming.
- [x] Make the bonus easier to get. Require less time behind for it to activate (1 second min should be enough) and allows a larger angle. Distance may be reduced slightly however.
- [x] Change the way the angle is calculated to compensate for kart steering.
- [x] Scale the bonus force with the time behind (scaling limited by caps). The bonus should overall be slightly smaller than before, making it less destabilizing ; the ease of activation more than compensating it.
- [x] Make the speed at which slipstream activate related to the usual max speed of the difficulty level, either by a fixed offset or by a relative percentage cap [OK with zone size scaling with speed]
- [x] Scale the slipstreaming zone and bonus with speed. This would not only be logical, but make it matter more in higher difficulties and less in lower. The speed at which it activates and all other characteristics could then be uniform without issues.
- [x] Make the AI understand slipstream better (make it use the bonus willingly) [The bonus scales so it is not necessary to use the bonus as soon as possible ; except if there is a danger. Further refinements are not in the scope of this issue.]

I think I can rework the non-GUI parts.

Here is a rough outline of what the GUI could look like : 
![slipstreammockup2](https://user-images.githubusercontent.com/25536748/37554124-aa60b98a-29d4-11e8-929e-201fc456e277.png)
![slipstreammockup](https://user-images.githubusercontent.com/25536748/37554127-aa9c535a-29d4-11e8-88e5-a2f8b4410438.png)

**Comment by Alayan-stk-2** [comment]
To solve this : "Change the way the angle is calculated to compensate for kart steering."

I think the easiest way is not to look at how the kart is steering, but to remember where he was a short time before (like 0,5 seconds before or the ticks equivalent). Then, simply compute the direction of the vector of the difference between the current position and the old position.

Running with --splistream-debug ; it's obvious that the current slipstream zone swings wildly whenever the kart changes orientation (even if its overall path is not that far from a straight line ; but with opposites skids) it should become much more stable with the aforementioned change.

**Comment by Alayan-stk-2** [comment]
"There is no visual indicator of slipstreaming progress"

Apparently, this is wrong. The code which displays the kind of white lines has a variable intensity depending on the slipstream time.

However, that I realized this by looking at the code and not in play proves that it is currently not clear enough.

Changing the color of the white lines to a yellow-orangish color when the bonus is ready would probably be a good idea.

In the same way, "Scale the bonus force with the time behind " is already done according to the detailed code : the slipstreaming time indicator is decremented when using the bonus, and it keeps renewing the bonus until it's zero. That is, the total bonus time = the time from the slipstreaming time indicator + the min bonus duration + (somewhat) the fadeout time.

This often wasn't obvious, especially with the previous speedometer max speed limitation ; but once aware of the rule, testing makes its existence quite clear. Still, some on/off light to say the bonus is active may be useful.

**Comment by Mrxx99** [comment]
"Changing the color of the white lines to a yellow-orangish color when the bonus is ready would probably be a good idea."

I had the exact same idea :D
But then it should be the same color as for skidding bonus, either the orange or the red

**Comment by Alayan-stk-2** [comment]
The main things have been fixed, but upon closer inspection (and with some time elapsed since I did the most painful changes to change the quad angle) ; some remaining issues have popped.

- [ ] No more visual effect when getting the bonus. Not a big issue when there is only one AI kart (the disappearance of the mesh is enough), but things become unclear when there are several karts.
- [ ] It may be possible to accumulate slipstreaming time from several karts at once (e.g. soon after race start when karts are close). This can make too easy to get the bonus
- [ ] With the larger zone, the timer decrease when going out of the zone works in a way that it's rather easy (read : a bit too much) to go trigger the bonus again with the next kart ahead.
